plugins/src/main/kotlin/org/gradle/kotlin/dsl/plugins/dsl/KotlinDslPlugin.kt

```diff
@@ -73,16 +84,57 @@ open class KotlinDslPlugin : Plugin<Project> {
     }
 
     private
-    fun Project.configureCompilerTasks() {
+    fun Project.enablePrecompiledScripts() {
         tasks {
-            "compileKotlin"(KotlinCompile::class) {
-                kotlinOptions {
-                    freeCompilerArgs += listOf("-script-templates", scriptTemplates)
+
+            val implicitImportsFile = kotlinDslBuildDir().map { it.file("implicit-imports") }
+            val generateImplicitImportsFile by creating {
+                inputs.property("gradleVersion", gradleVersion())
+                inputs.property("gradleKotlinDslVersion", gradleKotlinDslVersion())
+                outputs.file(implicitImportsFile)
+                doLast {
+                    writeImplicitImportsTo(implicitImportsFile)
                 }
             }
+
+            afterEvaluate {
+                "compileKotlin"(KotlinCompile::class) {
+                    dependsOn(generateImplicitImportsFile)
+                    kotlinOptions {
+                        freeCompilerArgs += listOf(
+                            "-script-templates", scriptTemplates,
+                            // Propagate implicit imports and other settings
+                            "-Xscript-resolver-environment=kotlinDslImplicitImportsFile=\"${implicitImportsFile.get().asFile.canonicalPath}\""
```

Comment:
:x: shouldn't the implicit import file be explicitly registered as a file input for the `KotlinCompile` task? we should also prove the changes in this PR don't break the cacheability of `KotlinCompile`